File: packages/python/elizaos/features/advanced_capabilities/actions/think.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING

from elizaos.generated.spec_helpers import require_action_spec
from elizaos.prompts import THINK_TEMPLATE
from elizaos.types import Action, ActionExample, ActionResult, Content, ModelType
from elizaos.utils.xml import parse_key_value_xml

if TYPE_CHECKING:
    from elizaos.types import HandlerCallback, HandlerOptions, IAgentRuntime, Memory, State

# Get text content from centralized specs
_spec = require_action_spec("THINK")


def _convert_spec_examples() -> list[list[ActionExample]]:
    """Convert spec examples to ActionExample format."""
    spec_examples = _spec.get("examples", [])
    if not isinstance(spec_examples, list):
        return []
    result: list[list[ActionExample]] = []
    for example in spec_examples:
        if not isinstance(example, list):
            continue
        row: list[ActionExample] = []
        for msg in example:
            if not isinstance(msg, dict):
                continue
            content = msg.get("content", {})
            text = ""
            actions: list[str] | None = None
            if isinstance(content, dict):
                text_val = content.get("text", "")
                text = str(text_val) if text_val else ""
                actions_val = content.get("actions")
                if isinstance(actions_val, list) and all(isinstance(a, str) for a in actions_val):
                    actions = list(actions_val)
            row.append(
                ActionExample(
                    name=str(msg.get("name", "")),
                    content=Content(text=text, actions=actions),
                )
            )
        if row:
            result.append(row)
    return result


@dataclass
class ThinkAction:
    """Deep thinking action that re-processes the full conversation context
    through a larger model when the initial planning pass determines the
    question needs deeper analysis."""

    name: str = _spec["name"]
    similes: list[str] = field(default_factory=lambda: list(_spec.get("similes", [])))
    description: str = _spec["description"]

    async def validate(
        self, runtime: IAgentRuntime, message: Memory, _state: State | None = None
    ) -> bool:
        return True

    async def handler(
        self,
        runtime: IAgentRuntime,
        message: Memory,
        state: State | None = None,
        options: HandlerOptions | None = None,
        callback: HandlerCallback | None = None,
        responses: list[Memory] | None = None,
    ) -> ActionResult:
        if state is None:
            raise ValueError("State is required for THINK action")

        # Compose full state with all available context
        state = await runtime.compose_state(message, ["RECENT_MESSAGES", "ACTION_STATE"])

        template = (
            runtime.character.templates.get("thinkTemplate")
            if runtime.character.templates and "thinkTemplate" in runtime.character.templates
            else THINK_TEMPLATE
        )
        prompt = runtime.compose_prompt(state=state, template=template)

        # Use the large model for deeper reasoning — this is the core
        # upgrade over the default planning pass which uses ACTION_PLANNER
        response = await runtime.use_model(ModelType.TEXT_LARGE, prompt=prompt)
        parsed_xml = parse_key_value_xml(response)

        if parsed_xml is None:
            raise ValueError("Failed to parse XML response for think action")

        thought = str(parsed_xml.get("thought", ""))
        text = str(parsed_xml.get("text", ""))

        if callback:
            await callback(Content(text=text, thought=thought, actions=["THINK"]))

        # The result flows to subsequent actions via previousResults.
        # Downstream actions see this as the first link in the chain.
        return ActionResult(
            text=text,
            values={
                "success": True,
                "responded": True,
                "lastReply": text,
                "thoughtProcess": thought,
            },
            data={
                "actionName": "THINK",
                "responseThought": thought,
                "responseText": text,
                "thought": thought,
                "messageGenerated": True,
            },
            success=True,
        )

    @property
    def examples(self) -> list[list[ActionExample]]:
        return _convert_spec_examples()


think_action = Action(
    name=ThinkAction.name,
    similes=ThinkAction().similes,
    description=ThinkAction.description,
    validate=ThinkAction().validate,
    handler=ThinkAction().handler,
    examples=ThinkAction().examples,
)

File: packages/rust/src/features/advanced_capabilities/actions/think.rs
//! THINK action implementation.

use async_trait::async_trait;
use once_cell::sync::Lazy;
use std::sync::Arc;

use crate::basic_capabilities::runtime::{IAgentRuntime, ModelParams};
use crate::error::{PluginError, PluginResult};
use crate::generated::spec_helpers::require_action_spec;
use crate::prompts::THINK_TEMPLATE;
use crate::types::{ActionResult, Memory, ModelType, State};
use crate::xml::parse_key_value_xml;

use super::Action;

// Get text content from centralized specs
static SPEC: Lazy<&'static crate::generated::spec_helpers::ActionDoc> =
    Lazy::new(|| require_action_spec("THINK"));

/// Action for deep thinking and careful reasoning.
/// Re-processes the full conversation context through a larger model when
/// the initial planning pass determines the question needs deeper analysis.
pub struct ThinkAction;

#[async_trait]
impl Action for ThinkAction {
    fn name(&self) -> &'static str {
        &SPEC.name
    }

    fn similes(&self) -> &[&'static str] {
        static SIMILES: Lazy<Box<[&'static str]>> = Lazy::new(|| {
            SPEC.similes
                .iter()
                .map(|s| s.as_str())
                .collect::<Vec<_>>()
                .into_boxed_slice()
        });
        &SIMILES
    }

    fn description(&self) -> &'static str {
        &SPEC.description
    }

    async fn validate(&self, _runtime: &dyn IAgentRuntime, _message: &Memory) -> bool {
        true
    }

    async fn handler(
        &self,
        runtime: Arc<dyn IAgentRuntime>,
        message: &Memory,
        state: Option<&State>,
        _responses: Option<&[Memory]>,
    ) -> PluginResult<ActionResult> {
        let _state = state.ok_or_else(|| {
            PluginError::StateRequired("State is required for THINK action".to_string())
        })?;

        // Compose full state with all available context
        let composed_state = runtime
            .compose_state(message, &["RECENT_MESSAGES", "ACTION_STATE"])
            .await?;

        let template = runtime
            .character()
            .templates
            .get("thinkTemplate")
            .map(|s| s.as_str())
            .unwrap_or(THINK_TEMPLATE);

        let prompt = runtime.compose_prompt(&composed_state, template);

        // Use the large model for deeper reasoning — this is the core
        // upgrade over the default planning pass which uses ACTION_PLANNER
        let response = runtime
            .use_model(ModelType::TextLarge, ModelParams::with_prompt(&prompt))
            .await
            .map_err(|e| PluginError::ModelError(e.to_string()))?;

        let response_text = response
            .as_text()
            .ok_or_else(|| PluginError::ModelError("Expected text response".to_string()))?;

        let parsed = parse_key_value_xml(response_text)
            .ok_or_else(|| PluginError::XmlParse("Failed to parse think response".to_string()))?;

        let thought = parsed.get("thought").cloned().unwrap_or_default();
        let text = parsed.get("text").cloned().unwrap_or_default();

        // The result flows to subsequent actions via previousResults.
        // Downstream actions see this as the first link in the chain.
        Ok(
            ActionResult::success(text.clone())
                .with_value("success", true)
                .with_value("responded", true)
                .with_value("lastReply", text.clone())
                .with_value("thoughtProcess", thought.clone())
                .with_data("actionName", "THINK")
                .with_data("responseThought", thought.clone())
                .with_data("responseText", text)
                .with_data("thought", thought)
                .with_data("messageGenerated", true),
        )
    }
}
